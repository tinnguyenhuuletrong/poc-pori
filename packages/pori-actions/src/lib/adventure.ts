import { Context, PromiseReturnType } from '@pori-and-friends/pori-metadata';
import { random } from 'lodash';

const ALL_SLOTS = [1, 2, 3, 4, 5, 6, 7, 8, 9];

export function randAdventureSlot(
  samples: number,
  excludeIndex: number[] = []
) {
  let pool = ALL_SLOTS.filter((itm) => !excludeIndex.includes(itm));

  const res = [];
  for (let i = 0; i < samples; i++) {
    const next = pool[random(0, pool.length - 1, false)];
    if (!next) throw new Error('not enough pool');

    pool = pool.filter((v) => v !== next);
    res.push(next);
  }

  return res;
}

export type SCMineInfo = PromiseReturnType<
  ReturnType<typeof queryMineinfoFromSc>
>;
export async function queryMineinfoFromSc(ctx: Context, mineId: number) {
  const res = await ctx.contract.methods.mines(mineId - 1).call();
  const farmer = parseMinePlayer(ctx, res.farmer);
  const helper = parseMinePlayer(ctx, res.helper);
  const rewardMap = parseRewardMap(ctx, res.rewardMap);
  return {
    farmer,
    helper,
    rewardMap,
  };
}

function parseMinePlayer(ctx: Context, playerInfo: any) {
  const address = playerInfo.player;
  const selectedIndex = ctx.web3.utils
    .hexToBytes(playerInfo.selectedCells)
    .filter((itm) => itm > 0);
  return {
    address,
    selectedIndex,
  };
}

function parseRewardMap(ctx: Context, rawRewardMap: string) {
  const bytes = ctx.web3.utils.hexToBytes(rawRewardMap);
  const env = byte2number(bytes.slice(0, 2));
  const startTimeUnixSec = byte2number(bytes.slice(20, 28));

  const slots: Record<string, { reward: number; joined: number }> = {};
  const startOffset = 2;
  for (let i = 0; i < 9; i++) {
    const reward = bytes[startOffset + i * 2];
    const joined = bytes[startOffset + i * 2 + 1];
    slots[i] = { reward, joined };
  }

  return {
    env,
    startTimeUnixSec,
    startTimeInDate: new Date(startTimeUnixSec * 1000),
    slots,
    mineRawRewadMap: rawRewardMap,
  };
}

function byte2number(bytes: number[]) {
  return parseInt(Buffer.from(bytes).toString('hex'), 16);
}

export async function queryRandomRewardLevelFromSc(
  ctx: Context,
  mineInfo: SCMineInfo
) {
  return ctx.contract.methods
    .randomRewardLevel(mineInfo.rewardMap.mineRawRewadMap)
    .call();
}

export async function queryMissiontOfPoriSc(
  ctx: Context,
  pori: string | number
) {
  const engagedMission = await ctx.contract.methods.missionOfPori(pori).call();
  return parseInt(engagedMission);
}

export type SCPortalInfo = PromiseReturnType<
  ReturnType<typeof queryPortalInfoSc>
>;
export async function queryPortalInfoSc(ctx: Context, addr: string) {
  const info = await ctx.contractPortal.methods.portalInfoOf(addr).call();
  const { missions, fastMissions, capacityMissions } = info;
  return {
    missions: parseInt(missions),
    fastMissions: parseInt(fastMissions),
    capacityMissions: parseInt(capacityMissions),
  };
}
