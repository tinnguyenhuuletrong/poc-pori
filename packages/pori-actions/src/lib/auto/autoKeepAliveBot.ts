import { Workflow } from '../../index';
import { Context } from '@pori-and-friends/pori-metadata';
import { AutoPlayDb, captureStartedBot, takeABreak } from './autoPlayWorkflow';

export type StartBotFuncHandler = () => Promise<void>;
export type AutoKeepAliveBotArgs = {
  type: 'keep_alive_bot';
  intervalMs: number;
  botDb: Record<string, StartBotFuncHandler>;
};

export async function autoKeepAliveBot({
  ctx,
  realm,
  args,
}: {
  ctx: Context;
  realm: Realm;
  args: AutoKeepAliveBotArgs;
}) {
  const intervalMs = args.intervalMs;
  const botId = `keep_alive_bot`;
  if (AutoPlayDb[botId]) {
    ctx.ui.writeMessage(`bot with id ${botId} is running. skip it`);
    return;
  }

  const workflowExec = async (state: Workflow.WorkflowState) => {
    let count = 0;
    state.updateState(() => {
      state.data['_it'] = count;
      state.data['_nextAt'] = new Date(Date.now() + intervalMs);
    });

    // eslint-disable-next-line no-constant-condition
    while (true) {
      await takeABreak(state, ctx, intervalMs);

      for (const id in args.botDb) {
        if (AutoPlayDb[id]) continue;

        ctx.ui.writeMessage(`${botId} restart bot ${id}`);
        await args.botDb[id]();
      }

      state.updateState(() => {
        count++;
        state.data['_it'] = count;
        state.data['_nextAt'] = new Date(Date.now() + intervalMs);
      });
    }
  };

  const state = Workflow.createWorkflow(workflowExec, botId);
  state
    .start()
    .catch((err) => {
      ctx.ui.writeMessage(
        `keepAliveBot #bot${state.id} error ${err.toString()}`
      );
    })
    .finally(() => {
      ctx.ui.writeMessage(`keepAliveBot #bot${state.id} end!`);
    });

  ctx.ui.writeMessage(
    `keepAliveBot #bot${state.id} started:
    - Interval: ${intervalMs / (1 * 60 * 1000)} mins
    `
  );

  captureStartedBot(state, args);
  return state;
}
