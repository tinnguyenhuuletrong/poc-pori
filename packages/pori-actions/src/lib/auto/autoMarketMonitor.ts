import { queryMarketItems } from '../queryPoriApi';
import { Workflow } from '../../index';
import {
  Context,
  getMarketplayBaseLink,
  NFTItemPotionIds,
  NFTItemSeedIds,
  TEN_POWER_10_BN,
} from '@pori-and-friends/pori-metadata';
import { isEmpty } from 'lodash';
import { AutoPlayDb, captureStartedBot, takeABreak } from './autoPlayWorkflow';

export type AutoMonitorMarketItemPriceArgs = {
  type: 'market_items_monitor';
  intervalMs: number;
  minSeedToNotice: number;
  minPotionToNotice: number;
};

export async function autoMonitorMarketItemPrices({
  ctx,
  realm,
  args,
}: {
  ctx: Context;
  realm: Realm;
  args: AutoMonitorMarketItemPriceArgs;
}) {
  const intervalMs = args.intervalMs;
  const botId = `market_items_monitor`;
  if (AutoPlayDb[botId]) {
    ctx.ui.writeMessage(`bot with id ${botId} is running. skip it`);
    return;
  }

  const workflowExec = async (state: Workflow.WorkflowState) => {
    let count = 0;
    state.updateState(() => {
      state.data['_it'] = count;
      state.data['_nextAt'] = new Date(Date.now() + intervalMs);
    });

    // eslint-disable-next-line no-constant-condition
    while (true) {
      await takeABreak(state, ctx, intervalMs);

      await checkMarketItems(ctx, args, state);

      state.updateState(() => {
        count++;
        state.data['_it'] = count;
        state.data['_nextAt'] = new Date(Date.now() + intervalMs);
      });
    }
  };

  const state = Workflow.createWorkflow(workflowExec, botId);
  state
    .start()
    .catch((err) => {
      ctx.ui.writeMessage(
        `marketPriceMonitor #bot${state.id} error ${err.toString()}`
      );
    })
    .finally(() => {
      ctx.ui.writeMessage(`marketPriceMonitor #bot${state.id} end!`);
    });

  ctx.ui.writeMessage(
    `marketPriceMonitor #bot${state.id} started:
    - Interval: ${intervalMs / (1 * 60 * 1000)} mins
    `
  );

  captureStartedBot(state, args);
  return state;
}

async function checkMarketItems(
  ctx: Context,
  args: AutoMonitorMarketItemPriceArgs,
  state: Workflow.WorkflowState
) {
  const sellingItems = await queryMarketItems({ ctx, pageSize: 35 });
  const marketplaceBaseUrl = getMarketplayBaseLink(ctx.env);

  const seenItems = state.data['seenItems'] || {};

  const seedItems = sellingItems
    .filter((itm) => NFTItemSeedIds.includes(+itm.itemType.id))
    .filter((itm) => {
      const price = BigInt(itm.unitPrice) / TEN_POWER_10_BN;
      const isLowerThan = price <= args.minSeedToNotice;
      const hasSeen = !!seenItems[itm.id];

      return isLowerThan && !hasSeen;
    });

  const potionItems = sellingItems
    .filter((itm) => NFTItemPotionIds.includes(+itm.itemType.id))
    .filter((itm) => {
      const price = BigInt(itm.unitPrice) / TEN_POWER_10_BN;
      const isLowerThan = price <= args.minPotionToNotice;
      const hasSeen = !!seenItems[itm.id];

      return isLowerThan && !hasSeen;
    });

  const res = [...seedItems, ...potionItems];
  state.updateState(() => {
    res.forEach((itm) => (seenItems[itm.id] = true));
    state.data['seenItems'] = seenItems;
  });

  if (!isEmpty(res)) {
    const botMasterUid = process.env['TELEGRAM_MASTER_ID'];
    ctx.ui.writeMessage(
      `marketPriceMonitor #bot${state.id} notice:
      [master](tg://user?id=${botMasterUid}) ðŸŒœ items to check:

      ${res
        .map(
          (itm) =>
            `- orderId: ${itm.scOrderId} - ${itm.itemType.name} - price: ${
              BigInt(itm.unitPrice) / TEN_POWER_10_BN
            }`
        )
        .join('\n')}
      `
    );
  }
}
