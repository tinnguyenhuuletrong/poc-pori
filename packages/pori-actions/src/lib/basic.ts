import { EventEmitter } from 'stream';
import { Context } from '@pori-and-friends/pori-metadata';
import { BlockNumber } from 'web3-core';

export async function getMineInfo({ contract }: Context, mineId: string) {
  const mineInfo = await contract.methods.mines(mineId).call();
  const { state, farmer, helper, rewardMap } = mineInfo;
  return { state, farmer, helper, rewardMap };
}

export async function listenEvents({
  contract,
}: Context): Promise<EventEmitter> {
  // To scan pass event from blockA -> blockB
  // use
  //  https://web3js.readthedocs.io/en/v1.7.3/web3-eth-contract.html#getpastevents

  // Scan all event from currentTime
  return contract.events
    .allEvents()
    .on('connected', function (subscriptionId: string) {
      console.log(subscriptionId);
    });
}

export async function scanEvents(
  ctx: Context,
  {
    filter = 'allEvents',
    fromBlock,
    toBlock,
  }: {
    filter: string;
    fromBlock: BlockNumber;
    toBlock: BlockNumber;
  }
) {
  const contract = ctx.contract;

  return contract.getPastEvents(filter, {
    fromBlock,
    toBlock,
  });
}
