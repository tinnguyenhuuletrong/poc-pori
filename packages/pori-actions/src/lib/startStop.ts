import {
  Context,
  ENV,
  getIdleGameAddressSC,
  getWeb3NodeUri,
  getWeb3NodeUriHttp,
} from '@pori-and-friends/pori-metadata';
import { EventEmitter } from 'stream';
import Web3 from 'web3';

export async function init(env: ENV): Promise<Context> {
  const uriws = getWeb3NodeUri(env);
  const urihttp = getWeb3NodeUriHttp(env);

  if (!uriws && !urihttp) {
    console.error(`missing env NODE_URI_${env} or NODE_URI_${env}_HTTP`);
    process.exit(1);
  }

  const provider = uriws
    ? new Web3.providers.WebsocketProvider(uriws)
    : new Web3.providers.HttpProvider(urihttp);

  console.log(
    'use web3 provider',
    provider instanceof Web3.providers.WebsocketProvider ? 'wss' : 'http'
  );

  const web3 = new Web3(provider);

  const idleGameSc = getIdleGameAddressSC(env);
  const contract = new web3.eth.Contract(idleGameSc.abi, idleGameSc.address);

  const ctx: Context = {
    contract,
    web3,
    provider,
    env,
    emiter: new EventEmitter(),
  };

  return ctx;
}

export async function close(ctx: Context) {
  ctx.provider.disconnect();
}
