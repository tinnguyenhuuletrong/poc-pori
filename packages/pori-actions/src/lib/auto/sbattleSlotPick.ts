import { Adventure } from '../../index';
import { AdventureInfoEx, Context } from '@pori-and-friends/pori-metadata';
import { minBy, sortBy, sum, toString, uniq } from 'lodash';
import { ESB_P_THRESHOLD_KEEP_BIG_REWARD } from './autoPlayWorkflow';
import { minIndex, minIndexBy } from '@pori-and-friends/utils';
import { SCSCellInfo } from '../adventure';

/*
  SBattle
  1. assit 3 pories
    2MaxFarmerPower > 2MaxSupportPower => null

    // supporter have only 1 pori
    //  farmer may have 
    //    1 atk lower => need send more 
    //    0 => need send more
    
    no S => pick 1 pori rewardlevel min
    has S => pick 1 pori. atk > x AND rewardLevel min
  else 
  2. query 
      Adventure.getPoriansAtSCellSc -> {sFarmer, sHelper}
      2MaxFarmerPower, 2MaxSupportPower
  3. if esb(Sum(2MaxFarmerPower) , Sum(2MaxSupportPower)) < 15 => null

  [optional] - optimize cost
  3.5 find optimize pair in case 2MaxFarmerPower > 2MaxSupportPower
        example: 57525
        need to find (max1,max2) 
            - sum(max1,max2) > 2MaxSupportPower
            - min(rewardLevel(max1) + rewardLevel(max2))

  4. Return
     srcIds = [max1, max2].filter(id => !sFarmer.includes(id))
     desIds = sFarmer.filter(id => ![max1, max2].include(id))
     if empty(desIds)
        desIds.fill(srcIds.length,0)
     

swap tables
[3,4] [3]   => src [4] des[0]
[3,4] [4]   => src [3] des[0]
[3,4] [3,5] => src [4] des[5]
[3,4] []    => src [3,4] des [0,0]
[3,4] [5,6] => src [3,4] des [5,6]


3.5 note 
SumMaxEnemyPower
minCost = int.Max

for i 0 -> n-1
  for j i+i -> n
    i.p + j.p > SumMaxEnemyPower
      AND i.rewardLevel + j.rewardLevel < minCost
        
*/

type ParseMineInfo = ReturnType<typeof _parseMineInfo>;
export type SBattlerDecision = {
  missionId: number;
  srcIds: number[];
  desIds: number[];
  sTreasureIndex: number;
};
export async function sbattleSlotPick({
  mineInfo,
  sCellInfo,
  isFarmer,
  ctx,
}: {
  mineInfo: AdventureInfoEx;
  sCellInfo: SCSCellInfo;
  isFarmer: boolean;
  ctx: Context;
}): Promise<SBattlerDecision | null> {
  const parseMineInfo = _parseMineInfo(mineInfo);
  const {
    supporterPories,
    hasS,
    mineId,
    farmerMaxPowerOf2,
    supporterMaxPowerOf2,
    supporterPoriesSortedByDecPower,
    farmerPoriesSortedByDecPower,
    powerOf,
  } = parseMineInfo;
  if (!hasS) {
    await ctx.ui.writeMessage(`sbattle ${mineId}: s not found`);
    return null;
  }
  if (supporterPories.length <= 3) {
    await ctx.ui.writeMessage(`sbattle ${mineId}: assit have 3 pories case`);
    return sbattleSlotPickCase3({ mineInfo, sCellInfo, isFarmer, ctx });
  }

  if (supporterMaxPowerOf2 > farmerMaxPowerOf2) {
    // TODO: try to calculate ebs -> rely on lucky
    await ctx.ui.writeMessage(
      `sbattle ${mineId}: max2Power lesser than supporter ${farmerMaxPowerOf2} < ${supporterMaxPowerOf2}. do nothing for now. consider ebs lucky later`
    );
    return null;
  }

  // TODO: 4 pories case
  return null;
}

async function sbattleSlotPickCase3({
  mineInfo,
  sCellInfo,
  isFarmer,
  ctx,
}: {
  mineInfo: AdventureInfoEx;
  sCellInfo: SCSCellInfo;
  isFarmer: boolean;
  ctx: Context;
}): Promise<SBattlerDecision | null> {
  const parseMineInfo = _parseMineInfo(mineInfo);
  const { mineId, sIndex, powerOf } = parseMineInfo;
  const sPowerFarmer = sum(sCellInfo.farmer.map((itm) => powerOf(itm)));
  const sPowerSupporter = sum(sCellInfo.helper.map((itm) => powerOf(itm)));

  // no need to play. win S already
  if (sPowerFarmer > sPowerSupporter) {
    await ctx.ui.writeMessage(
      `sbattle ${mineId}: explorer power higher ${sPowerFarmer} vs ${sPowerSupporter} and assit can not play s. nothing to do`
    );
    return null;
  }

  const minAtk = powerOf(sCellInfo.helper[0]);
  const poriInfo = _pickOneMinRewardLevelAndAtkGtXPori(parseMineInfo, minAtk);
  if (!poriInfo) {
    await ctx.ui.writeMessage(
      `sbattle ${mineId}: don't have pori atk > ${minAtk}. nothing to do`
    );
    return null;
  }

  await ctx.ui.writeMessage(
    `sbattle ${mineId}: move min rewardLevel -> S. ${poriInfo.id} - R:${poriInfo.rewardLevel} - Pw:${poriInfo.power}`
  );
  return {
    missionId: mineId,
    srcIds: [poriInfo.id],
    desIds: [0],
    sTreasureIndex: sIndex,
  };
}

function _parseMineInfo(mineInfo: AdventureInfoEx) {
  const farmerPories = mineInfo.farmerPories || [];
  const farmerRewardLevel = mineInfo.farmerRewardLevel || [];
  const farmerSlots = mineInfo.farmerSlots || [];
  const supporterPories = mineInfo.supporterPories || [];
  const supporterRewardLevel = mineInfo.supporterRewardLevel || [];
  const supporterSlots = mineInfo.supporterSlots || [];
  const mineId = mineInfo.mineId;

  const activeIndexs = [
    ...(mineInfo?.farmerSlots || []),
    ...(mineInfo?.supporterSlots || []),
  ];
  const activeRewardLevels = [...farmerRewardLevel, ...supporterRewardLevel];
  const sIndex = activeIndexs[activeRewardLevels.indexOf(4)];
  const powerOf = (id: number | string) => {
    return mineInfo.powers[id.toString()];
  };

  const farmerPoriesSortedByDecPower = sortBy(
    farmerPories.map((itm, index) => ({
      id: itm,
      index,
      power: powerOf(itm),
      rewardLevel: farmerRewardLevel[index],
    })),
    (a) => a.power
  ).reverse();
  const supporterPoriesSortedByDecPower = sortBy(
    supporterPories.map((itm, index) => ({
      id: itm,
      index,
      power: powerOf(itm),
      rewardLevel: supporterRewardLevel[index],
    })),
    (a) => a.power
  ).reverse();

  const farmerMaxPowerOf2 =
    farmerPoriesSortedByDecPower[0].power +
    farmerPoriesSortedByDecPower[1].power;

  const supporterMaxPowerOf2 =
    farmerPoriesSortedByDecPower[0].power +
    farmerPoriesSortedByDecPower[1].power;

  return {
    mineId,
    sIndex,
    hasS: !!sIndex,
    farmerPories,
    farmerRewardLevel,
    farmerSlots,
    supporterPories,
    supporterRewardLevel,
    supporterSlots,

    farmerPoriesSortedByDecPower,
    supporterPoriesSortedByDecPower,
    farmerMaxPowerOf2,
    supporterMaxPowerOf2,
    powerOf,
  };
}

function _pickOneMinRewardLevelPori(mineInfo: ParseMineInfo) {
  const { farmerPoriesSortedByDecPower } = mineInfo;
  const res = minBy(farmerPoriesSortedByDecPower, (itm) => itm.rewardLevel);

  if (!res) return null;
  return res.id;
}

function _pickOneMinRewardLevelAndAtkGtXPori(
  mineInfo: ParseMineInfo,
  minAtk: number
) {
  const { farmerPoriesSortedByDecPower } = mineInfo;
  const tmp = farmerPoriesSortedByDecPower.filter((itm) => itm.power > minAtk);
  const res = minBy(tmp, (itm) => itm.rewardLevel);
  if (!res) return null;
  return res;
}
