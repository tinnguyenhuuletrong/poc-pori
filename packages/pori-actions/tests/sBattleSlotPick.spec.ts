import ld from 'lodash';
import { AdventureInfoEx, Context, ENV } from '@pori-and-friends/pori-metadata';

import { sbattleSlotPick } from '../src/lib/auto/sbattleSlotPick';
import { init } from '../src/lib/startStop';
import { SCSCellInfo } from '../src/lib/adventure';

type ContextEx = Context & {
  _logs?: string[];
};

describe('s_battle_slot_pick', () => {
  let ctx: ContextEx;
  const BASE_MINE_INFO: AdventureInfoEx = {
    mineId: 36430,

    isFarmer: true,
    farmerAddress: '0xBFFB3aC9305C9426b456e3a5C9dB45bee0bFF3aC',

    isSupporter: false,
    supporterAddress: '0xBFFB3aC9305C9426b456e3a5C9dB45bee0bFF3aC',

    hasBigReward: true,

    farmerPories: [225, 3830, 3560, 2257],
    farmerRewardLevel: [3, 1, 2, 3],
    farmerSlots: [3, 1, 7, 9],

    supporterPories: [5748, 5567, 5749],
    supporterRewardLevel: [2, 2, 1],
    supporterSlots: [2, 4, 8],
    link: '',
    blockedTo: new Date('2022-07-06T04:56:45.000Z'),
    atkAt: new Date(' 2022-07-06T02:26:45.000Z'),
    powers: {
      '225': 311,
      '2257': 318,
      '3560': 312,
      '3830': 312,
      '5567': 303,
      '5748': 308,
      '5749': 314,
    },
  };

  beforeAll(async () => {
    ctx = await init(ENV.ProdPorichain);
    ctx._logs = [];
    ctx.ui.writeMessage = async (msg) => {
      console.log(msg);
      ctx._logs?.push(msg);
    };
  });

  beforeEach(() => {
    ctx._logs = [];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('assit has 3p - not s', async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 2, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749],
      supporterRewardLevel: [2, 2, 1],
      supporterSlots: [2, 4, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 314,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: [],
      helper: [],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test('assit has 3p - has s. explorer power higher', async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749],
      supporterRewardLevel: [2, 2, 4],
      supporterSlots: [2, 4, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 300,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test('assit has 3p - has s. farmer 1 supporter 0', async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749],
      supporterRewardLevel: [2, 2, 1],
      supporterSlots: [2, 4, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 300,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: [],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test('assit has 3p - has s. farmer 1 supporter 1(higher)', async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749],
      supporterRewardLevel: [2, 2, 4],
      supporterSlots: [2, 4, 9],

      powers: {
        '225': 311,
        '2257': 318,
        '3560': 330,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 320,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 3p - has s. farmer 1 supporter 1(higher) - need swap 2 to win `, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749],
      supporterRewardLevel: [2, 2, 4],
      supporterSlots: [2, 4, 9],

      powers: {
        '225': 311,
        '2257': 318,
        '3560': 310,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 320,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 3p - mine power lower. but can swap 2`, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [100508, 5387, 6060, 6266],
      farmerRewardLevel: [2, 2, 4, 1],
      farmerSlots: [3, 8, 5, 9],

      supporterPories: [100766, 5131, 5425],
      supporterRewardLevel: [4, 3, 3],
      supporterSlots: [5, 7, 2],

      powers: {
        '5131': 314,
        '5387': 298,
        '5425': 311,
        '6060': 302,
        '6266': 305,
        '100508': 304,
        '100766': 322,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['5387'],
      helper: ['100766'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 4p - has s. farmer can win. p1 in s, p2 not`, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749, 5421],
      supporterRewardLevel: [2, 2, 4, 1],
      supporterSlots: [2, 4, 9, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3560': 310,
        '3830': 312,
        '5567': 303,
        '5748': 308,
        '5749': 300,
        '5421': 200,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 4p - has s. farmer can win. p1 p2 not s`, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 1, 1],
      farmerSlots: [3, 9, 7],

      supporterPories: [5748, 5567, 5749, 5421],
      supporterRewardLevel: [2, 2, 4, 1],
      supporterSlots: [2, 4, 9, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3560': 310,
        '3830': 280,
        '5567': 303,
        '5748': 308,
        '5749': 300,
        '5421': 200,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 4p - has s. farmer can win. p1 p2 not s 2`, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 4, 1],
      farmerSlots: [3, 9, 9],

      supporterPories: [5748, 5567, 5749, 5421],
      supporterRewardLevel: [2, 2, 4, 1],
      supporterSlots: [2, 4, 9, 1],

      powers: {
        '225': 311,
        '2257': 318,
        '3560': 270,
        '3830': 280,
        '5567': 303,
        '5748': 308,
        '5749': 300,
        '5421': 200,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830', '3560'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });

  test(`assit has 4p - has s. farmer can not win`, async () => {
    const mineInfo: AdventureInfoEx = {
      ...BASE_MINE_INFO,

      farmerPories: [225, 3830, 3560, 2257],
      farmerRewardLevel: [3, 4, 4, 1],
      farmerSlots: [3, 9, 9],

      supporterPories: [5748, 5567, 5749, 5421],
      supporterRewardLevel: [2, 2, 4, 1],
      supporterSlots: [2, 4, 9, 1],

      powers: {
        '225': 211,
        '2257': 218,
        '3560': 270,
        '3830': 280,
        '5567': 303,
        '5748': 308,
        '5749': 300,
        '5421': 200,
      },
    };

    const sCellInfo: SCSCellInfo = {
      farmer: ['3830', '3560'],
      helper: ['5749'],
    };

    const sCmd = await sbattleSlotPick({
      mineInfo,
      sCellInfo,
      isFarmer: true,
      ctx,
    });
    console.log(sCmd);

    expect({ sCmd, logs: ctx._logs }).toMatchSnapshot();
  });
});
