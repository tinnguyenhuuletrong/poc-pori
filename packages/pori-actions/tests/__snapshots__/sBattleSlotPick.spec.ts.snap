// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`s_battle_slot_pick assit has 3p - has s. explorer power higher 1`] = `
Object {
  "logs": Array [
    "sbattle 36430: assit have 3 pories case",
    "sbattle 36430: explorer power higher 312 vs 300 and assit can not play s. nothing to do",
  ],
  "sCmd": null,
}
`;

exports[`s_battle_slot_pick assit has 3p - has s. farmer 1 supporter 0 1`] = `
Object {
  "logs": Array [
    "sbattle 36430: assit have 3 pories case",
    "sbattle 36430: explorer power higher 312 vs 0 and assit can not play s. nothing to do",
  ],
  "sCmd": null,
}
`;

exports[`s_battle_slot_pick assit has 3p - has s. farmer 1 supporter 1(higher) - don't have  1`] = `
Object {
  "logs": Array [
    "sbattle 36430: assit have 3 pories case",
    "sbattle 36430: don't have pori atk > 320. nothing to do",
  ],
  "sCmd": null,
}
`;

exports[`s_battle_slot_pick assit has 3p - has s. farmer 1 supporter 1(higher) 1`] = `
Object {
  "logs": Array [
    "sbattle 36430: assit have 3 pories case",
    "sbattle 36430: move min rewardLevel -> S. 3560 - R:1 - Pw:330",
  ],
  "sCmd": Object {
    "desIds": Array [
      0,
    ],
    "missionId": 36430,
    "sTreasureIndex": 9,
    "srcIds": Array [
      3560,
    ],
  },
}
`;

exports[`s_battle_slot_pick assit has 3p - not s 1`] = `
Object {
  "logs": Array [
    "sbattle 36430: s not found",
  ],
  "sCmd": null,
}
`;
